Add initial ARCHITECTURE document (#749)
rojo-rbx/rojo: ARCHITECTURE.md

# Architecture

Rojo is a rather large project with a bunch of moving parts. While it's not too complicated in practice, it tends to be overwhelming because it's a fair bit of Rust and not very clear where to begin.

This document is a "what the heck is going on" level view of Rojo and the codebase that's written to make it more reasonable to jump into something. It won't go too into depth on *how* something is done, but it will go into depth on *what* is being done.

## Overarching

> Rojo is divided into two main pieces: the server and the plugin. The server is what's ran on your computer (whether it be via the terminal or the visual studio code extension), with the plugin serving as its main client.

Reviewer: ```suggestion
Rojo is divided into two main pieces: the server and the plugin. The server is what's ran on your computer (whether it be via the terminal, the visual studio code extension, or a remote machine), with the plugin serving as its main client.
```
Perhaps worth mentioning that it can work over a remote connection too?

When serving a project, the server gathers data on all of the files in that project, puts it into a nice format, and then sends it to the plugin. Then, when something changes on the file system, it does the same thing for only the changed files and sends them to the plugin.

> When it receives a patch (whether it be the initial patch or any subsequent ones), the plugin reads through it and attempts to to apply it. Any sugar (the patch visualizer, as an example) happens on top of the patches received from the server.

Reviewer: ```suggestion
When the plugin receives a patch from the server (whether it be the initial patch or any subsequent ones), the plugin reads through the patch and attempts to to apply the changes described by it. Any sugar (the patch visualizer, as an example) happens on top of the patches received from the server.
```
Too many ambiguous uses of "it".

Reviewer: Also, this section is mildly misleading, as the initial patch is actually computed by the plugin and not received from the server. That's a bit of complexity that is explained in the plugin section, so I think keeping it simple up here is fine. The application process is the same, and that's what the focus is here.

Author: > as the initial patch is actually computed by the plugin and not received from the server

I sort of assumed we were sending an initial patch file because we compute one when a `ServeSession` is initialized in the CLI as part of building the `RojoTree`. I imagine this is just a semantics difference though: regardless of what's sent, the initial patch would have to be computed by the plugin to bring it to parity with the server.

These patches represent changes, additions, and removals to the Roblox tree that Rojo creates and manages.

When generating snapshots, files are 'transformed' into Roblox objects through what's referred to as the `snapshot middleware`. As an example, this middleware takes files named `init.lua` and transforms them into a `ModuleScript` bearing the name of the parent folder. It's also responsible for things like JSON models and `.rbxm`/`.rbxmx` models being turned into snapshottable trees.

Inquiring minds should look at `snapshot/mod.rs` and `snapshot_middleware` for a more detailed explanation.

Because snapshots are designed to be translated into Instances anyway, this system is also used by the `build` command to turn a Rojo project into a complete file. The backend for serializing a snapshot into a file is provided by `rbx-dom`, which is a different project.

### The Web Server

> Rojo uses a small web server to forward changes to the plugin. Once a patch is computed by the snapshot system, it's made available via the server's API. Then, the plugin requests regularly, and if a new patch exists, recieves it and applies it in Studio.

Reviewer: ```suggestion
Rojo uses a small web server to forward changes to the plugin. Once a patch is computed by the snapshot system, it's made available via the server's API. Then, the plugin requests patches regularly using a technique called [long polling](https://en.wikipedia.org/wiki/Push_technology#Long_polling) to receive the next patch from the server and then applies it to the Datamodel in Studio.
```

The plugin doesn't poll, it long polls. This means the patch always exists from the plugin's perspective since it just waits until the patch does exist.

The web server itself is very basic, consisting of around half a dozen endpoints. The bulk of the work is performed by either the snapshot system or the plugin, with the web server acting as a middleman.

## The Plugin

> This section of the document is left incomplete.

Reviewer: I would love some rough stubs of things you'd like me to document here. (Can be as replies here/discord if you don't want to muddy the doc.)
For example, I don't think documenting the props of every basic component is needed, but documenting the methods on App is needed. I'd appreciate input on where to draw these lines and what info/sections would be most valuable.

Reviewer: Following up
We discussed externally, and the plan is to only document the plugin behavior that directly relates to the core functionality of Rojo, such as the reconciler and servesession. Things like patch visualization, undo/redo protection, etc will be documented elsewhere.

> This section of the document is left incomplete.

Reviewer: ```suggestion
As mentioned above, the Studio plugin uses the web server to receive patches that contain the changes, additions, and removals to the Studio datamodel. The patch does not actually point at instances directly but rather uses string IDs. The plugin has an InstanceMap that is a bidirectional map between instance IDs and Roblox instances. It lets us keep track of every instance that Rojo knows about.

To actually apply the patch to the Studio datamodel, the plugin has a Reconciler. The Reconciler uses the InstanceMap to find the instances mentioned by the patch, and then it attempts to modify/create/destroy instances as specified in the patch. It will also create and return a new patch that contains all of the operations that failed to apply (effectively creating a patch that if applied, would fix the failures), which is used to display warnings to the end user.

The first time the plugin connects to the server is a bit of a special case. The server gives the plugin the current state, which is first used to hydrate the InstanceMap with all the initially known instances. Then, the Reconciler computes a patch by diffing the datamodel with the server state. This is known as the "catch up" patch, and that is what is shown in the Confirming page in the plugin. When that catch up patch is applied, the Studio datamodel should match the Rojo server.
```

Author: I would prefer that this be left as a PR on [my branch](https://github.com/Dekkonot/rojo/tree/ted) rather than a suggestion here, since then it can be properly reviewed.

Reviewer: Moved it here: https://github.com/Dekkonot/rojo/pull/2

## Data Structures

> Rojo has many data structures and their purpose might not be immediately clear at a glance. To alleviate this, they are documented below.

Reviewer: We should also document the structures in the plugin, like PatchSet. It's particularly important because of how inconsistent things are with camelCase and PascalCase in our structures across the project.

Reviewer: We'll need to do this with Plugin and Server headers in this section due to the duplicate naming (ie: ServeSession) making the distinctions critical.

Reviewer: Following up
Because this doc is just meant as a high level introduction to Rojo functionality, we don't need to document the details of every plugin structure. I do think PatchSet is still very important to document though, because its the main thing the plugin interacts with during core functionality.

Author: We have a dedicated `Plugin` subsection that this can go in if need be, though maybe it's worth giving the Plugin its own document?
